### ISSUE TITLE
Nested facets

### ISSUE THREAD

**Issue by ivan-kleshnin** [Yomguithereal/baobab#226]
Another post-practical wish of mine. Currently facets are set in a flat list (object). It requires to **invent** new variable names where common namespacing could suffice. We need to deal with plural vs singular, peek right words... etc. Naming is among the hardest CS tasks, as you perfectly know.

Currently:
```js
facets: {
  currentRobot: { // custom name
    ...
  },
  currentMonster: { // custom name
    ...
  },
  currentRobots: { // custom name
    ...
  },
  currentMonsters: { // custom name
    ...
  },
}
```

Better
```js
facets: {
  robot: { // model name: namespace
    currentModel: { // synthetic name
      ...
    },
    currentIndex: {
      ...
    },
  },
  monster: { // model name: namespace
    currentModel: {
      ...
    },
    currentIndex: {
      ...
    }
  },
}
```

The current situation as I see it: **the more information you can express as a facet – the better**.
Facets are god-sent. Reactive wins.

**Comment by RJ** [comment]
I came to post a similar request for nested facets.

Here's my situation. Connections contain lists of buffers. Buffers are categorised into active, inactive, archived, etc.

<pre>
{
 buffers: {
  1: {...},
  2: {...},
  3: {...},
  ...
 },
 connections: {
  connectionOne: {
   active_buffers: [1,2],
   inactive_buffers: [3]    
  },
  connectionTwo: {
   archived_buffers: [7,8,9...]
  }
}
</pre>

I want to easily map over the buffer objects for each connection. I think it'd be useful to be able to do something like this - i'll prefix with a $ to denote the field is a facet:

<pre>
{
 buffers: {
  1: {...},
  2: {...},
  3: {...},
  ...
 },
 connections: {
  connectionOne: {
   active_buffers: [1,2],
   inactive_buffers: [3],
   $activeBuffers: activeBuffersFacet       <------- THIS
   $activeBuffers: buffersFromIdListFacet(this.active_buffers)     <------- or even this
  },
  connectionTwo: {
   archived_buffers: [7,8,9...]
  }
}
</pre>

Basically a way to join my somewhat-normalised data so the buffer objects appear as a map in the correct place in my tree. Whether or not that's defined in a separate facets section that mirrors the nesting structure of the tree is another matter, but that's my use case.

**Comment by Yomguithereal** [comment]
This could be interesting indeed to be able to nest the facets. I'll probably add this in an upcoming version.

**Comment by Yomguithereal** [comment]
In-tree facets are now supported in v2.

**Comment by ivan-kleshnin** [comment]
Seems not working.

> In-tree facets are now supported in v2.

In-tree or **nested**?

**Comment by Yomguithereal** [comment]
In-tree facets now working. But since you can place them wherever you want in the tree now, do we need nesting?

**Comment by ivan-kleshnin** [comment]
I'm not completely sure what do you mean by **in-tree**. 
The problem, which I've described in the first post (above) still applies.

You can't group your data
```js
{ // NOT WORKING
  robots: {
    $currentModel: { 
      ...
    },
    $currentModels: {
      ...
    },
  },
}
```
therefore you can't 

1) normally edit the code (fold / unfold specific sections, etc.).
2) pick semantically better names 

```js
robots: {
  $hasPendingRequest: ...
}
```

must be expressed as unreadable

```js
$hasPendingRequestRobot ?
$robotsHavePendingRequest ?
```

**Comment by Yomguithereal** [comment]
In v2 you can do the following:

```js
const tree = new Baobab({
  robots: {
     data: [],
     $currentModel: {...},
     $currentModels: {...},
     $hasPendingRequest: {...}
  }
});
```

Or nest event more than that if needed.

**Comment by ivan-kleshnin** [comment]
This is what I call **nesting**, and it didn't work for me. 
I'm going to make bug-proof or find my own error then.

**Comment by Yomguithereal** [comment]
If this does not work with v2 for you then yes, we do have a bug.

**Comment by ivan-kleshnin** [comment]
With this declaration

```js
robots: {
  ...
  $havePendingRequests: [
    ["ajaxQueue"],
    function (queue) {
      return ajaxQueueContains(queue, robotApi.indexUrl);
    }
  ],
},
```

I'm getting this error
**Baobab.Facet: attempting to create a computed data node with a wrong definition (path: /robots/$havePendingRequests)**

I wasn't able to reproduce it in a sandbox environment but I don't see any significant difference between one code base and another.

**Comment by Yomguithereal** [comment]
Ok, here what I tried:

```js
var Baobab = require('./src/baobab.js');

var tree = new Baobab({
  ajaxQueue: 'hey',
  robots: {
    $havePendingRequests: [
      ['ajaxQueue'],
      function(queue) {
        return queue + 'ho';
      }
    ]
  }
});

console.log(tree.get('robots'))
```

This works as expected.

The error you get tells us that the given facet definition is incorrect. Normally, the throw Error object should have a definition key. Can you tell me what it tells?

**Comment by ivan-kleshnin** [comment]
In normal cases this definition is something like `[Array[1], Array[1], function]`.
In error case it is `false`. There is some concurrency going there which prevents me from sandboxing and emulating it. I'll continue to experiment.

**Comment by ivan-kleshnin** [comment]
Opened https://github.com/Yomguithereal/baobab/issues/305
